vm/src/dictdatatype.rs

```diff
@@ -0,0 +1,216 @@
+/// Inspired by: https://morepypy.blogspot.com/2015/01/faster-more-memory-efficient-and-more.html
+/// And: https://www.youtube.com/watch?v=p33CVV29OG8
+/// And: http://code.activestate.com/recipes/578375/
+use std::collections::HashMap;
+
+pub struct Dict {
+    size: usize,
+    indices: HashMap<usize, usize>,
+    entries: Vec<Option<DictEntry>>,
+}
+
+struct DictEntry {
+    hash: usize,
+    key: PyObjectRef,
+    value: PyObjectRef,
+}
+
+impl Dict {
+    pub fn new() -> Self {
+        Dict {
+            size: 0,
+            indices: HashMap::new(),
+            entries: Vec::new(),
+        }
+    }
+
+    /// Store a key
+    pub fn insert(
+        &mut self,
+        vm: &mut VirtualMachine,
+        key: &PyObjectRef,
+        value: PyObjectRef,
+    ) -> Result<(), PyObjectRef> {
+        match self.lookup(vm, key)? {
+            LookupResult::Existing(index) => {
+                // Update existing key
+                if let Some(ref mut entry) = self.entries[index] {
+                    entry.value = value;
+                    Ok(())
+                } else {
+                    panic!("Lookup returned invalied index into entries!");
+                }
+            }
+            LookupResult::NewIndex { i, hash_value } => {
+                // New key:
+                let entry = DictEntry {
+                    hash: hash_value,
+                    key: key.clone(),
+                    value,
+                };
+                let index = self.entries.len();
+                self.entries.push(Some(entry));
+                self.indices.insert(i, index);
+                Ok(())
+            }
+        }
+    }
+
+    pub fn contains(
+        &self,
+        vm: &mut VirtualMachine,
+        key: &PyObjectRef,
+    ) -> Result<bool, PyObjectRef> {
+        if let LookupResult::Existing(index) = self.lookup(vm, key)? {
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Retrieve a key
+    pub fn get(&self, vm: &mut VirtualMachine, key: &PyObjectRef) -> PyResult {
+        if let LookupResult::Existing(index) = self.lookup(vm, key)? {
+            if let Some(entry) = &self.entries[index] {
+                Ok(entry.value.clone())
+            } else {
+                panic!("Lookup returned invalied index into entries!");
+            }
+        } else {
+            let key_repr = vm.to_pystr(key)?;
+            Err(vm.new_value_error(format!("Key not found: {}", key_repr)))
+        }
+    }
+
+    /// Delete a key
+    pub fn delete(
+        &mut self,
+        vm: &mut VirtualMachine,
+        key: &PyObjectRef,
+    ) -> Result<(), PyObjectRef> {
+        if let LookupResult::Existing(index) = self.lookup(vm, key)? {
+            self.entries[index] = None;
+            Ok(())
+        } else {
+            let key_repr = vm.to_pystr(key)?;
+            Err(vm.new_value_error(format!("Key not found: {}", key_repr)))
+        }
+    }
+
+    pub fn len(&self) -> usize {
```

Comment:
I think len should probably be O(1). Should we track the number of entries in the dictionary?

Reply:
I fully agree.